620 factor out lobby connection

Wesmania opened the pull request:
Move lobby-server-talking code outside _clientwindow.py as much as possible. There's still a conflation of client state and server connection state (client.state), but we won't touch if for now to avoid breaking it.

- [x] PR branch should be named `issuenum`-`fix`/`feature`/`cleanup`-`description`
- [x] Code is split into logical commits
- [x] Code has tests
- [x] Final commit includes "Fixes #issue" in commit message

When all builds pass and a maintainer is happy with your PR, the "ready" label will be applied. Please complete these tasks then:

- [x] Rebase onto develop
- [x] Add changelog entry
- [ ] Remove this entire template section

Wesmania committed:
Make a stub lobby connection class

Signed-off-by: Igor Kotrasinski <ikotrasinsk@gmail.com>

Wesmania committed:
Modify _clientwindow and related classes to use lobby connection

Signed-off-by: Igor Kotrasinski <ikotrasinsk@gmail.com>

Wesmania committed:
Move trivial message handlers to connection class

Signed-off-by: Igor Kotrasinski <ikotrasinsk@gmail.com>

Wesmania committed:
Move handle_game_info to connection class

Signed-off-by: Igor Kotrasinski <ikotrasinsk@gmail.com>

Wesmania committed:
Remove some apparently unnecessary bits from ClientState

Signed-off-by: Igor Kotrasinski <ikotrasinsk@gmail.com>

Wesmania committed:
Factor out talking to server itself from LobbyConnection

Signed-off-by: Igor Kotrasinski <ikotrasinsk@gmail.com>

Wesmania committed:
Factor out dispatch to an explicit dispatcher

Signed-off-by: Igor Kotrasinski <ikotrasinsk@gmail.com>

Wesmania committed:
Move trivial dispatch forwards to _clientwindow

Signed-off-by: Igor Kotrasinski <ikotrasinsk@gmail.com>

Wesmania committed:
Move client state back to _clientwindow

Refactoring showed that client state should indeed be handled by the
client rather than the connection to the lobby server.

Signed-off-by: Igor Kotrasinski <ikotrasinsk@gmail.com>

Wesmania committed:
Move send method to lobby_connection

Signed-off-by: Igor Kotrasinski <ikotrasinsk@gmail.com>

Wesmania committed:
Eliminate last dependency of LobbyConnection on the client

Signed-off-by: Igor Kotrasinski <ikotrasinsk@gmail.com>

Wesmania committed:
Have replays widget explicity use dispatcher for server messages

Signed-off-by: Igor Kotrasinski <ikotrasinsk@gmail.com>

Wesmania committed:
Rename LobbyConnection to LobbyInfo

After refactors the only thing LobbyConnection does is listening for
and emitting info events from the lobby server. Rename accordingly.

Signed-off-by: Igor Kotrasinski <ikotrasinsk@gmail.com>

Wesmania committed:
Factor out reconnecting code to separate class

Signed-off-by: Wesmania <ikotrasinsk@gmail.com>

Wesmania committed:
Add keepalive functionality to lobby reconnector

Signed-off-by: Wesmania <ikotrasinsk@gmail.com>

muellni commented on src/client/connection.py:
```diff
@@ -31,6 +31,31 @@ def __init__(self, connection):
+
+    @keepalive.setter
+    def keepalive(self, value):
+        self._keepalive = self.value
```
typo: `self._keepalive = value`

Wesmania committed:
Allow to enable and disable the lobby reconnecter

Signed-off-by: Wesmania <ikotrasinsk@gmail.com>

Wesmania committed:
Rework lobby connection to mirror connection status better

Subscribe to socket status directly, so connection state always reflects
whether we're connected. Remove 'DROPPED' state from lobby connection
since 'DISCONNECTED' is functionally identical.

Signed-off-by: Wesmania <ikotrasinsk@gmail.com>

Wesmania committed:
Fix client not relogging in at reconnect

Signed-off-by: Igor Kotrasinski <ikotrasinsk@gmail.com>

Wesmania committed:
Clean up ClientState

Clean up the mishmash of states, make it driven only by connection
and login state. Remove unused auth_state, clean up logo use of
ClientState.

Signed-off-by: Igor Kotrasinski <ikotrasinsk@gmail.com>

Wesmania committed:
Remove unused variable in LobbyConnection

Signed-off-by: Igor Kotrasinski <ikotrasinsk@gmail.com>

Wesmania committed:
Add changelog entry for #620

Fixes #620.

Signed-off-by: Igor Kotrasinski <ikotrasinsk@gmail.com>
